src/liballoc/collections/vec_deque.rs

```diff
@@ -2163,19 +2183,40 @@ impl<'a, T> Iterator for Iter<'a, T> {
     }
 
     fn fold<Acc, F>(self, mut accum: Acc, mut f: F) -> Acc
-        where F: FnMut(Acc, Self::Item) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
     {
         let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
         accum = front.iter().fold(accum, &mut f);
         back.iter().fold(accum, &mut f)
     }
 
-    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R where
-        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Ok = B>,
     {
         let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        let accum = front.iter().try_fold(init, &mut f)?;
-        back.iter().try_fold(accum, &mut f)
+        let mut front_iter = front.iter();
+        let res = front_iter.try_fold(init, &mut f);
+        // the slice's `try_fold` leaves the iterator where we want our `tail` to be
+        self.tail = unsafe {
+            front_iter
+                .as_slice()
+                .as_ptr()
+                .offset_from(self.ring.as_ptr()) as usize
+        };
```

Comment:
I'm debating whether the unsafe is worth it here.  I think, using the same idea you have here, it could be something like `wrap_add(self.tail + (front.len() - front_iter.as_slice().len()), self.ring.len())`.

Reply:
Oh, actually, `offset_from` panics on ZSTs, so I think we can't use it here.

Reply:
Your suggestion looks too complex, but the ZST problem is a real one. I'll think of a better solution.

Reply:
After thinking a bit more about it, I changed my mind. Using `len` to calculate the correct offset is fine.